enhancements/scheduling/priority-class.md

```diff
@@ -0,0 +1,110 @@
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in
+  [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+By default OpenShift has two reserved priority classes for critical system pods
+to have guaranteed scheduling, this proposal suggests the need for a third less
+critical priority class which would be used for user facing features we ship
+with OpenShift, for example user workload monitoring pods, that are not system
+critical but should have a higher scheduling priority than any user workloads.
+
+## Motivation
+
+Currently OpenShift monitoring ships with two workloads, user and cluster
+monitoring. These include prometheus pods, they both have the same priority
+classes set. What can and actually did happen, is that with not enough
+resources, user workload prometheus pods were scheduled in favour of the
+cluster monitoring prometheus pods. This should not happen, as cluster
+monitoring is system critical component, whereas user monitoring has less
+priority. 
+
+### Goals
+
+1. Describe the current state of pod priority classes in OpenShift.
+2. Introduce a new third priority class to be used within OpenShift by any non
+   critical workloads.
+
+### Non-Goals
+
+Out of scope of this proposal is solving user specific priority classes.
+
+## Proposal
+
+Priority indicates the importance of a Pod relative to other Pods. If a Pod
+cannot be scheduled, the scheduler tries to preempt (evict) lower priority Pods
+to make scheduling of the pending Pod possible.
+
+By default OpenShift has two reserved priority classes for critical system pods
+to have guaranteed scheduling: (following is from our docs [1]):
+- `system-cluster-critical` - This priority class has a value of 2000000000
+  (two billion) and is used with pods that are important for the cluster. Pods
+  with this priority class can be evicted from a node in certain circumstances.
+  For example, pods configured with the system-node-critical priority class can
+  take priority. However, this priority class does ensure guaranteed
+  scheduling.  Examples of pods that can have this priority class are fluentd,
+  add-on components like descheduler, and so forth.
+- `system-node-critical` - This priority class has a value of 2000001000 and is
+  used for all pods that should never be evicted from a node. Examples of pods
+  that have this priority class are sdn-ovs, sdn, and so forth.
+
+As mentioned before these two are not enough and this proposal suggests
+introducing a third priority class: `user-critical`. This would be used by any
+pods that are important for user facing OpenShift features but are not deemed
+system critical. Example of pods include user workload monitoring and user's
+Knative Service. 
+
+### Risks and Mitigations
+
+N/A
+
+## Design Details
+
+New priority class would be created by the component that creates the two
```

Comment:
Thanks for the suggestions!
1. Adding a prefix of `openshift-` to the `user-critical` class name would make it clear to users that this is reserved for user workloads that are managed by OpenShift, but it goes against the current pattern as no one of our existing classes have a reserved prefix. To change this would require to change approx 170+ instances of the existing class names. It might make sense here as we would name it `user-critical` to prefix with `openshift-user-critical`, so I am happy to go that route just for this one.
2. Seems like we do not limit the scope of namespaces for the current two existing classes, but this should be an easier fix to do and I think we should do this for all 3 (two existing and the new proposeed). The pros for this are that users don't use and abuse this reserved class. The cons are we might be breaking some things for our users that already consume this class? But also potentially some of our Red Hat components that are not installed in openshift-* namespaces, are we okay with doing that?

 My vote would be to not change the existing classes, but apply the above to the new class, as we do not know what it might impact. Sounds good to you?

Will add to the document once we agree on this.

Reply:
@lilic I am confused by your comment.

Kubernetes reserves the system-* prefix for Kubernetes usage (see: https://github.com/kubernetes/kubernetes/blob/5238e1c80fea02891e1804b346d24faa7c13da07/pkg/apis/scheduling/validation/validation.go#L37)

The question I was asking is if OpenShift should reserve a similar openshift-* prefix for names that are unique to the distribution.  We obviously shouldn't change names that are reserved upstream.

Reply:
I am inclined to reserve the openshift-* name prefix for use by OpenShift.

Reply:
thinking on this more, i wonder if operator-framework may want to define some priority class bands under the

`operator-framework-*`prefix.

This would let us have a common (upstream oriented channel) for us to maybe align knative/istio/etc on maybe more standardized priority bands.  OLM itself could then take over ownership of these priority classes.

@bbrowning @ecordell @jwforres thoughts?

Reply:
@derekwaynecarr i think @lilic 's point was we have "system-*" today.  If we were starting fresh we'd make it "openshift-system-*".

So her concern was if we want a user equivalent, do we start a new pattern ("openshift-user-*") which is inconsistent w/ those existing ones (system-*), or do we align w/ the existing pattern (user-*)?  (or do we rename the existing ones to add the openshift prefix which is clearly not viable)

it sounds like your preference would be to introduce this one new with an openshift prefix, not touch the existing ones, and live w/ the inconsistency.  that's certainly fine w/ me.

i'm less thrilled with "openshift-framework" but i'm willing to go w/ anything that moves this forward, i don't think the name is that important, particularly with an openshift prefix that makes it clear its for our usage.

Reply:
@derekwaynecarr I think the further upstream (openshift -> operator-framework -> kubernetes) we can define a prefix like this the more widely adopted it may become. I don't have a strong preference, but do agree that something around the operator-framework level may be a good home for this. There are likely a large number of operators with pods that are less critical than core Kubernetes pods but more critical than individual user workload pods.

@lilic Regarding restricting the usage of new priority classes to specific namespaces, I feel that depends mostly on the prefix chosen and the intention behind that prefix. A prefix reserved by kubernetes or operator framework should not have any namespace enforcement. An `openshift-*` prefix could have namespace enforcement, but we've already seen that there are optional operators in the OpenShift ecosystem that do not install into `openshift-*` namespaces.

I think the answer to both questions comes down to if this is an enhancement purely for OpenShift and specific optional operators controlled by OpenShift or if it's for a broader pattern of how operators with pods that should always be a higher priority than user workloads request that priority while remaining a lower priority than core Kubernetes system pods.

Reply:
@bparees we cant make system-* openshift-system-* because Kubernetes itself reserves system and manages the two defined priority classes.  The concept presented in this proposal is asking if OpenShift should define a third, and if so, should we reserve a prefix as part of that proposal (i think so), but I feel like operator-framework is a decent place to define some priority bands for common use cases to further mitigate required operator permissions so that RHACM, ServiceMesh, etc. can all benefit with a common priority banding.  Maybe that means we end up with 4 priority classes, but before proceeding on this, would like to see someone from operator framework weigh in.

Reply:
this proposal doesnt say which component manages the new priority class described.
is it monitoring?  is it the cvo?  is it olm?  etc.

Reply:
> @bparees we cant make system-* openshift-system-* because Kubernetes itself reserves system and manages the two defined priority classes. 

ok but we could introduce it and stop using system-* if we wanted to.  (again, not that i am suggesting we do so and nor was Lili)

> this proposal doesnt say which component manages the new priority class described.
is it monitoring? is it the cvo? is it olm? etc.

The proosal states: 
`New priority class would be created by the component that creates the two existing priority classes`

Now, given that it sounds like that is an upstream component, perhaps that's not reasonable here, but whatever the closest downstream equivalent is would make sense to me.  If we assume this is going to be static (as i assume the others are?) then the config-operator seems like a sensible place for it.

If we intend to let admins modify it, well we can still let the config-operator create it, with the create-only annotation.

If we go with your operator-framework- prefix suggestion, then having OLM create it makes sense to me.

Reply:
Sorry for the late reply, PTO, and busy times at the monitoring team! Great discussion here, thanks all!

Regarding `operator-framework-` prefix suggestion, monitoring stack is not installed or has anything to do with openshift-framework or OLM, so for our component that would not make that much sense to use, at least I don't see that, but agreed it should be added for serverless most likely, but happy to include that as part of this proposal but maybe as a 4th class as was mentioned.

Another thing to note is we are installed before OLM so if they create that class we could not make use of it, so we would prefer a class that is created by a component that is installed sooner than the monitoring one, so CVO seems like a good fit there. Preferably the class name would be `openshift-user-critical ` for example as to me that reflects this component is critical to openshift users but not to openshift plaftorm/system component, but I don't have that strong opinions and happy to adjust it.

@derekwaynecarr @bbrowning @bparees  So the two questions I guess are: 
1. Would `openshift-user-critical` be okay for monitoring user Prometheus to use created by CVO, for now I am sure other components will use it in the future?
2. Should we create `openshift-framework-` prefix for anything installed by OLM as part of this proposal?

Reply:
@lilic i'm on board with "openshfit-user-critical" (for use by things that are not olm, such as CMO).  @derekwaynecarr seems like the primary hold-out on this.

As for openshift-framework-, I don't know how we "create" the openshift-framework- prefix to reserve it, absent some actual specific priority classes we want created, but if someone wants to define those priority classes as well, i'm fine with that as well as a solution for things are *not* core openshift but still need additional priority classes.

Reply:
Yes agreed Ben. @derekwaynecarr let me know what we can adjust to either get this merged or come up with an alternative for our issue, would be nice to make use of this for monitoring stack in 4.7. Thanks!

Reply:
@derekwaynecarr friendly ping, if you get a chance do you mind having a look above, thanks!

Reply:
I am +1 on `openshift-user-critical` and reserving the `openshift-` for the distribution.  It can be used by components in our release payload, and optionally by any OLM delivered content.

Reply:
Now with naming agreed, we need to find a component that creates this. As discussed with Derek out of band, we think the best place would be openshift-apiserver or openshift-controller-manager.

@mfojtik @deads2k any objections to proposed components or preference or even other ideas on where this pod priority class might be created? Thanks!

Reply:
>  As discussed with Derek out of band, we think the best place would be openshift-apiserver or openshift-controller-manager.

was cluster-config-operator explicilty rejected in that conversation?

Reply:
These are not OpenShift-API related. So neither oa nor ocm make sense. I also tend to cluster-config-operator.

Reply:
Thanks for the reply! The problem was not all components install cluster-config-operator, so would be a problem if a core component like monitoring uses this role but it's not there in some environments.

Reply:
@lilic if you get stuck on this i suggest putting it in the CMO itself for now and re-homing it when someone else wants to use it.

Reply:
@bparees Sounds great, will do, thanks!